Adding generation of TypeScript definitions for the angular code.

alexkrauss opened the pull request:
This pull request continues the idea of the `ts` branch, by adding the ability to generate .d.ts files for the generated angular code. In particular, types are generated for the parameters and results of all methods, which allows interacting with an API in a statically typed fashion, including IDE support like code completion.

See the test for a simple example.

Also merged master into the ts branch.

wcandillon commented:
Woah very cool. Is it possible to add a nodejs-typedef as well?

alexkrauss commented:
@wcandillon I haven't looked at the node code yet, but I suppose it should be easy to add for this variant. Probably, the only difference will be the type of the returned promise...

wcandillon commented:
@alexkrauss  yes, it's just about changing the return type.

alexkrauss commented:
OK. Then we can simply make this another template variable. Will try to do this tonight or tomorrow.

alexkrauss commented:
OK, as of 1ca4d3107e9f I added the node variant as well, as well as the possibility to add references to other typescript definitions (which is needed to resolve angular or node types).

Note that the result type `Q.Promise<{response: http.IncomingMessage, body: ...}>` for the node code just arised from me staring at the generated code and doing type inference in my head. I won't have time to actually write some typescript code against the generated stuff, so I cannot really test this right now. So I marked all this as experimental.

wcandillon commented on README.md:
```diff
@@ -246,6 +246,26 @@ eg: https://cloud.google.com/appengine/docs/go/requests#Go_Request_headers
+var CodeGen = require('swagger-js-codegen').CodeGen;
+
+var file = 'swagger/spec.json';
+var swagger = JSON.parse(fs.readFileSync(file, 'UTF-8'));
```
Can you add an example with `getTypeScriptDefinition` as well?

alexkrauss commented:
The interface is really the same, so I just added a little note. Also removed the `dtsRefs`, which can actually be covered by the general `mustache` option.

douglas-treadwell commented:
Is anything happening with this PR?

alexkrauss commented:
@wcandillon Any chance of merging and releasing this change any time soon? I'd like to use this on a project where dependencies to unreleased versions from a git repository lead to technical and social challenges :-)

If I can help in any way, let me know.

wcandillon commented:
@alexkrauss thks for this great contribution. As you will see, I have used your code partially for the typescript support.
